egs/wsj/s5/steps/data/reverberate_data_dir.py

```diff
@@ -0,0 +1,352 @@
+                        "--noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> "
+                        "--bg-fg-type <choices=(background|foreground), default=background> "
+                        "--rir-file <str, compulsary if isotropic, should not be specified if point-source> "
+                        "< location=(support Kaldi IO strings) >")
+    parser.add_argument("--num-replications", type=int, dest = "num_replicas", default = 1,
+                        help="Number of replicate to generated for the data")
+    parser.add_argument('--foreground-snrs', type=str, dest = "foreground_snr_string", default = '20:10:0', help='snrs for foreground noises')
+    parser.add_argument('--background-snrs', type=str, dest = "background_snr_string", default = '20:10:0', help='snrs for background noises')
+    parser.add_argument('--prefix', type=str, default = None, help='prefix for the id of the corrupted utterances')
+    parser.add_argument("--speech-rvb-probability", type=float, default = 0.8,
+                        help="Probability of reverberating a speech signal, e.g. 0 <= p <= 1")
+    parser.add_argument("--noise-adding-probability", type=float, default = 0.4,
+                        help="Probability of adding point-source noises, e.g. 0 <= p <= 1")
+    parser.add_argument("--max-noises-per-minute", type=int, default = 2,
+                        help="This controls the maximum number of point-source noises that could be added to a recording according to its duration")
+    parser.add_argument('--random-seed', type=int, default=0, help='seed to be used in the randomization of impulese and noises')
+    parser.add_argument("input_dir",
+                        help="Input data directory")
+    parser.add_argument("output_dir",
+                        help="Output data directory")
+
+    print(' '.join(sys.argv))
+
+    args = parser.parse_args()
+    args = CheckArgs(args)
+
+    return args
+
+def CheckArgs(args):
+    if not os.path.exists(args.output_dir):
+        os.makedirs(args.output_dir)
+
+    ## Check arguments.
+    if not os.path.isfile(args.rir_list_file):
+        raise Exception(args.rir_list_file + "not found")
+    
+    if args.noise_list_file is not None:
+        if not os.path.isfile(args.noise_list_file):
+            raise Exception(args.noise_list_file + "not found")
+
+    if args.num_replicas > 1 and args.prefix is None:
+        args.prefix = "rvb"
+        warnings.warn("--prefix is set to 'rvb' as --num-replications is larger than 1.")
+
+    return args
+
+
+# This function pick the item according to the associated probability
+# The input could be either a dictinoary of a list
+def PickItemWithProbability(x):
+   if isinstance(x, dict):
+     plist = list(set(x.values()))
+   else:
+     plist = x
+   total_p = sum(item.probability for item in plist)
+   p = random.uniform(0, total_p)
+   accumulate_p = 0
+   for item in plist:
```

Comment:
It's only calling x.values() once within that function, not more than once.

The only scenario there might be a problem is where an adversary with
knowledge of the random number generator is reordering the values in a
special way each time.  If the random numbers are truly random this
wouldn't be even possible.  Even if it were possible, it's paranoid.


On Tue, Jul 19, 2016 at 1:25 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r71416268>:
>
> > +        warnings.warn("--prefix is set to 'rvb' as --num-replications is larger than 1.")
> > +
> > +    return args
> > +
> > +
> > +# This function pick the item according to the associated probability
> > +# The input could be either a dictinoary of a list
> > +def PickItemWithProbability(x):
> > +   if isinstance(x, dict):
> > +     plist = list(set(x.values()))
> > +   else:
> > +     plist = x
> > +   total_p = sum(item.probability for item in plist)
> > +   p = random.uniform(0, total_p)
> > +   accumulate_p = 0
> > +   for item in plist:
>
> As the ordering of sets is not maintained across calls the inverse
> cumulative distribution function would change in each call.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/3802fde4b4da335c1482da6519f3f8d58145200f#r71416268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu2lpbSPHxbQ9Tdejs7Hko2IEevkFks5qXTKxgaJpZM4IJM2C>
> .
>